Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp

```diff
@@ -0,0 +1,303 @@
+        return;
+    }
+
+    queue().dispatchSync([weakThis = ThreadSafeWeakPtr { *this }, persistentScripts, completionHandler = WTFMove(completionHandler)]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis) {
+            completionHandler({ });
+            return;
+        }
+
+        String errorMessage;
+        if (!protectedThis->openDatabaseIfNecessary(errorMessage, true)) {
+            WorkQueue::mainSingleton().dispatch([errorMessage = crossThreadCopy(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
+                completionHandler(errorMessage);
+            });
+            return;
+        }
+
+        ASSERT(errorMessage.isEmpty());
+
+        for (const auto& script : persistentScripts)
+            protectedThis->insertScript(script, *(protectedThis->database()), errorMessage);
+
+        WorkQueue::mainSingleton().dispatch([errorMessage = crossThreadCopy(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
+            completionHandler(errorMessage);
+        });
+    });
+}
+
+void WebExtensionRegisteredScriptsSQLiteStore::getScripts(CompletionHandler<void(Vector<Ref<JSON::Object>> scripts, const String& errorMessage)>&& completionHandler)
+{
+    queue().dispatch([weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis) {
+            completionHandler({ }, nullString());
+            return;
+        }
+
+        String errorMessage;
+        auto scripts = protectedThis->getScriptsWithErrorMessage(errorMessage);
+        WorkQueue::mainSingleton().dispatch([scripts = WTFMove(scripts), errorMessage = crossThreadCopy(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
+            completionHandler(scripts, errorMessage);
+        });
+    });
+}
+
+Vector<Ref<JSON::Object>> WebExtensionRegisteredScriptsSQLiteStore::getScriptsWithErrorMessage(String& errorMessage)
+{
+    assertIsCurrent(queue());
+
+    if (!openDatabaseIfNecessary(errorMessage, false))
+        return { };
+
+    ASSERT(errorMessage.isEmpty());
+
+    if (RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT * FROM registered_scripts"_s))
+        return getKeysAndValuesFromRowIterator(*rows);
+    return { };
+}
+
+Vector<Ref<JSON::Object>> WebExtensionRegisteredScriptsSQLiteStore::getKeysAndValuesFromRowIterator(Ref<WebExtensionSQLiteRowEnumerator> rows)
+{
+    Vector<Ref<JSON::Object>> results;
+
+    RefPtr row = rows->next();
+    while (row != nullptr) {
+        auto script = row->getString(1);
+
+        auto value = JSON::Value::parseJSON(script);
+        if (RefPtr object = value->asObject())
+            results.append(*object);
+        else
+            RELEASE_LOG_ERROR(Extensions, "Failed to deserialize registered content scripts for extension %s", uniqueIdentifier().utf8().data());
```

Comment:
Need to null check the output of `parseJSON` and use `RefPtr`.
```suggestion
        if (RefPtr value = JSON::Value::parseJSON(script)) {
            if (RefPtr object = value->asObject())
                results.append(*object);
            else
                RELEASE_LOG_ERROR(Extensions, "Failed to deserialize registered content scripts for extension %s", uniqueIdentifier().utf8().data());
        } else
            RELEASE_LOG_ERROR(Extensions, "Failed to parse JSON for registered content scripts for extension %s", uniqueIdentifier().utf8().data());
```